doc/man/slogkey.1.in
.TH "slogkey"1 "" "" slogkey
.PP
.SS "Name"
.PP
slogkey — Manage cryptographic keys for use with the syslog-ng secure logging environment.
.SS "Synopsis"
.PP
slogkey [options] [arguments]
.SS "Description"
.PP
The slogkey utility is used to manage cryptographic keys for use with the secure logging module of syslog-ng. Use this utility to create a master key, derive a host key to be used by a secure logging configuration and to display the current sequence counter of a key. The options determine the operating mode and are mutually exclusive.
.SS "Arguments"
.PP
The arguments depend on the operating mode.
.RS
.IP * 2
Master key generation
.IP
Call sequence: slogkey --master-ḱey &lt;filename&gt;
.IP
&lt;filename&gt;: The name of the file to which the master key will be written.
.IP *
Host key derivation
.IP
Call sequence: slogkey --derive-key &lt;master key file&gt; &lt;host MAC address&gt; &lt;host serial number&gt; &lt;host key file&gt;
.RS
.IP * 2
&lt;master key file&gt;: The master key from which the host key will be derived.
.IP *
&lt;host MAC address&gt;: The MAC address of the host on which the key will be used. Instead of the MAC address, any other string that uniquely identifies a host can be supplied, e.g. the company inventory number.
.IP *
&lt;host serial number&gt;: The serial number of the host on which the key will be used. Instead of the serial number, any other string that uniquely identifies a host can be supplied, for example, the company inventory number.
.IP *
&lt;host key file&gt;: The name of the file to which the host key will be written.
.RE
.PP
NOTE: The newly created host key has its counter set to 0 indicating that it represents the initial host key k0. This host key must be kept secret and not be disclosed to third parties. It will be required to successfully decrypt and verify log archives processed by the secure logging environment. As each log entry will be encrypted with its own key, a new host key will be created after successful processing of a log entry and will replace the previous key. Therefore, the initial host key needs to be stored in a safe place before starting the secure logging environment, as it will be deleted from the log host after processing of the first log entry.
.IP *
Sequence counter display
.IP
Call sequence: slogkey --counter &lt;host key file&gt;
.RS
.IP * 2
&lt;host key file&gt;: The host key file from which the sequence will be read.
.RE
.RE
.SS "Options"
.RS
.IP * 2
--master-key or -m
.IP
Generates a mew master key. &lt;filename&gt; is the name of the file storing the newly generated master key.
.IP *
--derive-key or -d
.IP
Derive a host key using a previously generated master key.
.IP *
--counter or -c
.IP
Display the current log sequence counter of a key.
.IP *
--help or -h
.IP
Display a help message.
.RE
.SS "Files"
.PP
/usr/bin/slogkey
.PP
/etc/syslog-ng.conf
.SS "See also"
.PP
\fBsyslog-ng.conf.5\fP (\fIhttps://axoflow.com/docs/axosyslog-core/app-man-syslog-ng/syslog-ng.conf.5/\fP)
.PP
\fBsecure-logging.7\fP (\fIhttps://axoflow.com/docs/axosyslog-core/app-man-syslog-ng/secure-logging.7/\fP)
.SS "Getting help"
.RS
.IP * 2
The up-to-date documentation of syslog-ng is available on the \fBAxoSyslog documentation site\fP (\fIhttps://axoflow.com/docs/axosyslog-core/\fP).
.IP *
For news and notifications about syslog-ng, visit the \fBAxoflow blog\fP (\fIhttps://axoflow.com/blog/\fP).
.IP *
If you want to contact the developers directly to help with problems or report issues, contact us on \fBDiscord\fP (\fIhttps://discord.gg/583Z4wjem2\fP) or \fBGitHub\fP (\fIhttps://github.com/syslog-ng/syslog-ng/issues/\fP).
.RE
.PP
This manual page is maintained by \fBAxoflow\fP (\fIhttps://axoflow.com/\fP)
Last modified November 24, 2023: \fBMinor manpage fix (2f0a144)\fP (\fIhttps://github.com/axoflow/axosyslog-core-docs/commit/2f0a144ec391a3182da3af2f2cc136042d5608e9\fP)

NEWS
3.5.6
        Mon,  4 Aug 2014 13:39:42 +0200

        This is the sixth bug-fix release for the 3.5.x series.
        Upgrading from earlier versions is highly recommended, as the
        changes in this release are very small, yet, also very
        important for most platforms and workloads.

        Bugfixes
        ========

        * A major memory leak was fixed in the value-pairs framework,
          which affects $(format-json), MongoDB, AMQP and more. The
          leak was supposed to be fixed in 3.5.5, but due to a merging
          mistake, it was missed.

        * The configure script now detects Linux capabilities
          properly, there is no need to use --disable-linux-caps on
          non-Linux platforms anymore.

        * The pdbtool merge command will now generate version 4
          patterndb files, instead of version 3.

        Credits
        =======

        syslog-ng is developed as a community project, and as such it relies
        on volunteers to do the work necessary to produce syslog-ng.

        Reporting bugs, testing changes, writing code or simply providing
        feedback are all important contributions, so please if you are a
        user of syslog-ng, contribute.

        These people have helped in this release:

        Fabien Wernli
        Gergely Nagy
        Peter Czanik

3.5.5
        Mon, 21 Jul 2014 17:00:00 +0200

        This is the fifth bug-fix release for the 3.5.x series.
        Starting with this release, the 3.5 branch replaces 3.4 as the
        current "stable" version. Upgrading from earlier versions -
        and especially earlier branches - is highly recommended.

        New features
        ============

        * The facility() filter now supports ranges, allowing one to
          capture a range of facilities: facility(local0..local6).

        * The $(format-json) template function now supports SEQNUM.

        * The upstart script in contrib has been improved to allow the
          user to override certain settings (such as the configuration
          file, and the pidfile), and allow extra options to be
          specified.

        Bugfixes
        ========

        * Major memory leaks have been fixed in various parts of
          syslog-ng, including but not limited to network and file
          destinations and sources, value-pairs (thus affecting
          mongodb, $(format-json), and others), template handling and
          rewrite rules.

          Some of these leaks only happened during reload, some
          leaked on every message. They have now been plugged.

        * The value-pairs framework has seen a lot of updates, too:

          - It now honors time zone settings, so everything that uses
            value-pairs, will correctly take time zone settings into
            account.

          - Using globs in the key() option of value pairs will now
            also search macros for matches (thus key("D*") will match
            DATE too).

          - The syslog-proto (RFC5424) scope now correctly includes
            SDATA.

          - A bug that resulted in hierarchical elements being
            misparsed and made to appear in random places have been
            fixed.

        * The body() and subject() options of the SMTP destinations
          are now mandatory, and a double-free was also fixed in the
          driver.

        * Minor fixes were made to the $(env) template function, so
          that when looking up an empty environment variable, we will
          no longer trigger a glib warning. A misleading error message
          was also corrected.

        * The pdbtool patternize tool will now output version 4 PDB
          XML.

        * The configure script will not try to configure the AMQP
          module if no suitable version of python is installed. This
          avoids a difficult to diagnose error message during
          compilation.

        Developer news
        ==============

        * When compiled with internal ivykis, we will now install its
          headers too, along with our own. This is to make it easier
          for the Incubator to be compiled against a syslog-ng with
          embedded ivykis.

        * Plugin symbols are now loaded differently, in a way that
          lets plugins load further plugins, and use already loaded
          symbols. This is required for the various scripting
          destinations in the Incubator.

        * Starting with this release, we are installing the testing
          helpers as a convenience library (available under the
          syslog-ng-test pkg-config module). This allows third party
          projects to use our helpers.

        Credits
        =======

        syslog-ng is developed as a community project, and as such it relies
        on volunteers to do the work necessary to produce syslog-ng.

        Reporting bugs, testing changes, writing code or simply providing
        feedback are all important contributions, so please if you are a
        user of syslog-ng, contribute.

        These people have helped in this release:

        Balazs Scheidler <bazsi@balabit.hu>
        Daniel Gados
        Fabien Wernli
        Gergely Nagy <algernon@balabit.hu>
        Laszlo Budai <lbudai@balabit.hu>
        Laszlo Meszaros <lmesz@balabit.hu>
        Nick Alcock <nix@esperi.org.uk>
        Peter Czanik <czanik@balabit.hu>
        Peter Gyongyosi <gyp@balabit.hu>
        Sebastien Badia <seb@sebian.fr>
        Tibor Benke <tibor.benke@balabit.com>
        Viktor Juhasz <viktor.juhasz@balabit.com>
        Viktor Tusa <tusa@balabit.hu>
        Xufeng Zhang <xufeng.zhang@windriver.com>
        Zoltan Pallagi <pzolee@balabit.hu>

3.5.4.1
        Thu, 13 Mar 2014 10:46:07 +0100

        This is the fourth bug-fix release for the 3.5.x series, with
        a single fix for 32-bit platforms. Upgrading to this version
        is highly recommended for people using 32-bit platforms,
        others are not affected.


modules/secure-logging/slogkey/slogkey.c
/*
 * Copyright (c) 2019 Airbus Commercial Aircraft
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * As an additional exemption you are allowed to compile & link against the
 * OpenSSL libraries as published by the OpenSSL project. See the file
 * COPYING for details.
 *
 */

#include <stdio.h>
#include <sys/types.h>
#include <string.h>
#include <stdlib.h>

#include <glib.h>

#include <openssl/rand.h>
#include <openssl/evp.h>
#include <openssl/sha.h>

#include "messages.h"
#include "slog.h"

// Options
static gboolean master = FALSE;
static gboolean host = FALSE;
static gboolean counter = FALSE;

static GOptionEntry entries[] =
{
  { "master-key", 'm', 0, G_OPTION_ARG_NONE, &master, "Generate a master key", NULL },
  { "derive-host-key", 'd', 0, G_OPTION_ARG_NONE, &host, "Derive a host key from an existing master key", NULL },
  { "counter", 'c', 0, G_OPTION_ARG_NONE, &counter, "Display current host key counter", NULL },
  { NULL }
};

int main(int argc, char **argv)
{
  GError *error = NULL;
  GOptionContext *context = g_option_context_new("- secure logging key management\n\n  " \
                                                 "Master key generation:\tslogkey -m MASTERKEY\n  " \
                                                 "Host key derivation:\t\tslogkey -d MASTERKEY MACADDRESS SERIALNUMBER HOSTKEY\n  " \
                                                 "Host key counter display:\tslogkey -c HOSTKEY");

  g_option_context_add_main_entries (context, entries, NULL);
  if (!g_option_context_parse (context, &argc, &argv, &error))
    {
      g_print ("Invalid option: %s\n", error->message);
      exit (1);
    }

  if (argc < 2)
    {
      g_print("%s", g_option_context_get_help(context, TRUE, NULL));
      g_option_context_free(context);
      return -1;
    }

  gboolean ok = TRUE;

  // Options are mutually exclusive
  if (master && host)
    {
      ok = FALSE;
    }
  else if (master && counter)
    {
      ok = FALSE;
    }
  else if (host && counter)
    {
      ok = FALSE;
    }
  else if (master && host && counter)
    {
      ok = FALSE;
    }

  if (!ok)
    {
      g_print("%s", g_option_context_get_help(context, TRUE, NULL));
      g_option_context_free(context);
      return -1;
    }

  // Master key generation and sequence counter display need one option and a single argument
  if ((master || counter) && argc > 2)
    {
      g_print("%s", g_option_context_get_help(context, TRUE, NULL));
      g_option_context_free(context);
      return -1;
    }

  // Host key derivation needs one option and four arguments
  if(host && argc != 5)
    {
      g_print("%s", g_option_context_get_help(context, TRUE, NULL));
      g_option_context_free(context);
      return -1;
    }

  // Option parsing successful -> context can be released
  g_option_context_free(context);

  int ret = 0;
  int success = 0;
  int index = 1;

  // Initialize internal messaging
  msg_init(TRUE);

  if (master)
    {
      guchar masterkey[KEY_LENGTH];
      char *keyfile = argv[index];

      success = generateMasterKey(masterkey);
      if(!success)
        {
          msg_error("Unable to create master key");
          ret = -1;
        }
      else
        {
          success = writeKey((gchar *)masterkey, 0, keyfile);
          if(!success)
            {
              msg_error("[SLOG] ERROR: Unable to write master key to file", evt_tag_str("file", keyfile));
              ret = -1;
            }
        }
      return ret;
    }
  else if (counter)
    {
      // Display key counter
      char key[KEY_LENGTH];
      char *keyfile = argv[index];
      guint64 counterValue;
      success = readKey(key, &counterValue, keyfile);
      if(!success)
        {
          msg_error("[SLOG] ERROR: Unable to read key file", evt_tag_str("file", keyfile));
          return ret;
        }
      printf("counter=%zu\n", counterValue);
    }
  else if (host)
    {
      // Arguments
      gchar *masterKeyFileName = argv[index];
      index++;
      gchar *macAddr = argv[index];
      index++;
      gchar *serial = argv[index];
      index++;
      gchar *hostKeyFileName = argv[index];

      gchar masterKey[KEY_LENGTH] = { 0 };

      guint64 counterValue;

      success = readKey((char *)masterKey, &counterValue, masterKeyFileName);
      if (!success)
        {
          msg_error("[SLOG] ERROR: Unable to read master key", evt_tag_str("file", masterKeyFileName));
          ret = -1;
        }
      else
        {
          guchar hostKey[KEY_LENGTH];

          success = deriveHostKey((guchar *)masterKey, macAddr, serial, hostKey);
          if(!success)
            {
              msg_error("[SLOG] ERROR: Unable to derive a host key");
              ret = -1;
            }
          else
            {
              success = writeKey((char *)hostKey, 0, hostKeyFileName);
              if(success == 0)
                {
                  msg_error("[SLOG] ERROR: Unable to write host key to file", evt_tag_str("file", hostKeyFileName));
                  ret = -1;
                }
            }
        }
    }

  // Release messaging resources
  msg_deinit();

  return ret;
}
